### PR TITLE
Allow dropping dyn principal

### LINKED ISSUE #126313: `trait Trait: Send`: upcasting from `dyn Trait` to `dyn Send` does not work
The following compiles:
```rust
#![feature(trait_upcasting)]
trait Super {}
trait Trait: Super {}
fn foo(x: &dyn Trait) -> &dyn Super { x }
```

while this does not:
```rust
#![feature(trait_upcasting)]
trait Super {}
trait Trait: Send {}
fn foo(x: &dyn Trait) -> &dyn Send { x }
```

cc @compiler-errors I thought we actually had a PR for that at some point :thinking: I think that's actually not the case though. It wasn't part of #119338

### PR DESCRIPTION
**Pull request by WaffleLapkin**
Revival of #126660, which was a revival of #114679. Fixes #126313.

Allows dropping principal when coercing trait objects, e.g. `dyn Debug + Send` -> `dyn Send`.

cc @compiler-errors @Jules-Bertholet  
r? @lcnr

### CODE REVIEW

**Review by compiler-errors — APPROVED** [review]

File: tests/ui/traits/dyn-drop-principal.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+const fn yeet_principal(x: Box<dyn Any + Send>) -> Box<dyn Send> {
+    x
+}
+
+trait Bar: Send + Sync {}
+
+impl<T: Send + Sync> Bar for T {}
+
+const fn yeet_principal_2(x: Box<dyn Bar>) -> Box<dyn Send> {
```
Comment (compiler-errors) [inline tests/ui/traits/dyn-drop-principal.rs:14]: yay thanks for testing this one specifically :3

**Comment by compiler-errors** [comment]
r=me after ci passes, unless you want a review from lcnr specifically

File: tests/ui/traits/dyn-drop-principal.rs (hunk 1 above)
Comment (WaffleLapkin) [reply tests/ui/traits/dyn-drop-principal.rs:14]: This was actually @Jules-Bertholet :D
I only added a `dyn*` test as asked by lcnr.

Comment (compiler-errors) [reply tests/ui/traits/dyn-drop-principal.rs:14]: ty jules :3

**Pull request merged**